### ISSUE TITLE
Looking for someone smart who can figure out how to merge these backups

### ISSUE THREAD

**Issue by ShadyMedic** [bepaald/signalbackup-tools#151]
Okay fellas, so I found this awesome tool and successfully merged two backups together, although I might have screwed it up a little.

I had Signal on my old phone since I created my account and had all messages there. However, about a month ago, that phone bit the dust. I managed to get one of the latest Signal backups out of that, let's call it "Backup A".

On my laptop, I've been using Signal desktop for quite some time. I used it for the past two years or so, although I used Signal on my phone for much longer. So, a few years of the oldest messages are only in Backup A. Since my old phone stopped working, I continued to use this desktop installation of Signal. During that period, I started to message a few new people, whose threads are not present in Backup A. Let's call the data in my desktop version "Backup B".

What I did was merging Backup A with Backup B using the `--importfromdesktop` option. Then, I imported the resulting backup (let's call it "Backup AB") to my new phone, sent a "test" message to those few contacts that I started to correspond with on desktop only, so the threads are created in the Android Signal database. Then, I created a backup on the new phone (let's call it "Backup C"), transfered it back to my laptop and merged it with the desktop data once again. This time, the messages from the few new contacts located in desktop only were also included in the resulting backup. Let's call this "Backup BC".

I took this backup and imported it into the new phone, after I reinstalled Signal (so I get the option to restore from backup again).

The issue was, that most of the older messages were trippled. I suppose that this tool just appeneded the messages from the desktop installation (Backup B) to those already present in Backup A without checking for dupes. And then, when I created Backup C and merged it with Backup B again, the same thing happened, meaning that Backup BC had 3 coppies of messages that were present in Backup A and Backup B.

Before I realised this issue, I kept messaging with people. My desktop installation now has the proper chat history without duplicates, but misses a few years of my oldest messages. My android installation has all the messages, but thousands of them are trippled.

How would you go from here? I heard something about cropping Backup A to stop where Backup B started and then merge these two together, however, at the point where Backup B started (when I started using Signal desktop), I had only a few contacts in my Signal and if there is still the requirement that only threads present in the input backup (Backup A) can be merged with desktop data (Backup B), all of the more recent contacts/threads will not be present in the resulting backup.

I can try to draw a visualisation, if needed, but it's currently too late in the night for me here and I'm not sure if it's necessary.

Thanks a lot for this tool again and I'd be grateful if you could point me in the direction of a possible solution for this situation. If there is none, I have plan B, which is going through the entire chat histories and manually deleting the trippled messages, but that really isn't going to be fun.

**Comment by bepaald** [comment]
Hi!

>The issue was, that most of the older messages were trippled. I suppose that this tool just appeneded the messages from the desktop installation (Backup B) to those already present in Backup A without checking for dupes.

Yes, this is correct. Checking for dupes is not easy: the timestamps for the Desktop and Android messages will be different anyway. And message contents ('yes', '👍', or an empty message when sending a picture with no text) could be the same without messages being dupes. I do plan to add a `--removedoubles` option to deal with this, but it's not going to be exact: it _will_ delete messages that are not double and/or leave message that are.

The easiest way to deal with this is normally to add the `--limittodates` option when importing from desktop (see the readme, [here](https://github.com/bepaald/signalbackup-tools#desktop)). With the option to also crop the Android database to certain dates, the options to do this are almost endless, assuming you still have backup A. I'd say you should want to try to minimize the work done by this tool (in case it makes mistakes).

If you still have the original backup A, this is still what I would do:

1. Restore backup A, send some "test" messages so that all your newer contacts are in there, export a new backup A+.
2. Import the desktop (B) into A+, adding the `--limittodates` option. This option can take a list of date-ranges (again, see the readme) if you want to get fancy, but you could also just pass `"2019-09-18 00:00:00","2030-01-01 00:00:00"` (adjusting the first date of course), and have the "test" messages from step 1 still be doubled.
3. done! I think.

If you do not have backup A anymore, only B and BC, things get more complicated. You could still crop BC to the part that doesn't exist in C and then import (with limittodates) to get at least part of your messages undoubled, though the messages that were originally only in A will still be doubled.

>I can try to draw a visualisation, if needed, but it's currently too late in the night for me here and I'm not sure if it's necessary.

I _think_ I understood, but if some of the above doesn't make sense, I may not have.

>Thanks a lot for this tool again and I'd be grateful if you could point me in the direction of a possible solution for this situation. If there is none, I have plan B, which is going through the entire chat histories and manually deleting the trippled messages, but that really isn't going to be fun.

Yes I understand, I would recommend you then wait for me to implement the `--removedoubles` option (though it could take a good while!). I hope you still have backup A!

Let me know how it goes!

**Comment by bepaald** [comment]
>Yes, I still have Backup A. So I tried the following (in hopes it's what you suggested).
> [...]
>Did I do something wrong?

No, that all looks good. I'm not sure what is going on.

You did, in your output redact a crucial line. When I run this command on one of my testing backups, I get:
```
signalbackup-tools (./signalbackup-tools) source version 20231003.185421
BACKUPFILE VERSION: 1
BACKUPFILE SIZE: 65455705
COUNTER: 3011343483
Reading backup file...
FRAME 1804 (100.0%)... Read entire backup file...

done!
Database version: 209
  Using range: 1695199016903 - 1696954309756 (1695199016903 - 1696954309756)
Skipping conversation, conversation has no messages in requested time period (1/9)
Trying to match conversation (2/9) (type: private)
 - Importing 21 messages into thread._id 6
Skipping conversation, conversation has no messages in requested time period (3/9)
Skipping conversation, conversation has no messages in requested time period (4/9)
Trying to match conversation (5/9) (type: private)
 - Importing 7 messages into thread._id 9
Skipping conversation, conversation has no messages in requested time period (6/9)
Trying to match conversation (7/9) (type: group)
 - Importing 1 messages into thread._id 13
Skipping conversation, conversation has no messages in requested time period (8/9)
Trying to match conversation (9/9) (type: group)
 - Importing 1 messages into thread._id 16
reorderMmsSmsIds
updateThreadsEntries
  Dealing with thread id: 8, 1, 6, 5, 7, 9, 10, 13, 14, 15, 16
Checking foreign key constraints... ok
Checking database integrity (full)... ok
```

Note the line after `Database version: 209`, where it says 'Using range:', did you get that line, did it look correct?

I seem to recall from another issue maybe Windows screws up the numbers passed on the command line? If that could be the case, you could try the same thing, but adding quotes around the numbers: `--limittodates "1695199016903","1696954309756"`. If that is not it, I will have to think a bit about this.

Also, before trying again, if you would please update? I just added some code to hopefully fix the warnings you saw at the end (``Column `union_body' not found in query results``), I would like to know if these are gone in the current version.

Thanks!

**Comment by bepaald** [comment]
***OOPS***, missed one important thing in your message:

>6. Took SignalMergedBackup.backup and dumped it using signalbackup-tools_win.exe "SignalMergedBackup.backup" PASSPHRASE --importfromdesktop --output SignalDump --onlydb

You are _again_ adding `--importfromdesktop` here, so you are indeed doubling the messages again. To check if the 'SignalMergedBackup.backup' is ok, just `--output` it, do _not_ import from desktop again.

My guess is, SignalMergedBackup.backup is good. Let me know!

**Comment by ShadyMedic** [comment]
Wait, no, even after the update, I still see these warning as well, they were just kinda hidden in the threads list.

```
[...]
  Dealing with thread id: [REDACTED] 116Warning: Column `union_date' not found in query results
Warning: Column `union_body' not found in query results
Warning: Column `union_type' not found in query results
Warning: Column `mms._id' not found in query results
, [REDACTED]
[...]
```

I suppose it's not a big deal, but I guess you'd like to know about it.

**Comment by bepaald** [comment]
Excellent news! Happy to hear that!

>Yes, silly me forgot to remove the --importfromdesktop option. And when I was trying again a while ago, I forgot to replace the Backup A+ filename with the MergedBackup filename and was getting frustrated why the output file is the same lol.

Haha, no worries, those things will always happen. Just yesterday, when I was testing the `--limittodates` option after your last message, I was expecting a certain range of messages to be doubled, but not all of them were. I was already on an hour long bug-hunt before I realized the last messages in the range weren't doubled because they weren't in the Android backup to begin with (it was older). The single messages I was seeing were actually imported from the desktop correctly. 😵‍💫

>The only remaining warnings I see are:
>1. Warning: attachment data not found (rowid: 1??3, uniqueid: 16?????????09) – I suppose this is caused just by a deleted attachement, so not a big deal..?

Yeah, that's usually the case. 

> 2. Warning: Unsupported message type 'group-v2-change'. Skipping... (this warning will be shown only once) – No idea what does this mean.

The `group-v2-change` messages are when something about a group changes ("John changed the group name to 'something'", or "John added Jane to the group"). There are dozens of different sub-types of these messages and they are quite complicated in the Android database. Most if them are not supported to be imported. 

---

>I suppose it's not a big deal, but I guess you'd like to know about it.

Thanks, it is indeed not a big deal, but still good to know. I noticed I made a stupid typo in my previous fix, so hopefully _now_ it's fixed. 

Anyway, thanks for your kind words. I hope everything in your backup stays working now.

Cheers!
